pygeos/constructive.py

```diff
@@ -146,9 +152,21 @@ def buffer(
     True
     """
     if isinstance(cap_style, str):
-        cap_style = BufferCapStyles[cap_style.upper()].value
+        if not cap_style in VALID_BUFFER_CAP_STYLES:
+            raise ValueError(
+                "cap_style '{}' is not valid, must be one of {}".format(
+                    cap_style, ", ".join(VALID_BUFFER_CAP_STYLES)
+                )
+            )
+        cap_style = BufferCapStyles[cap_style].value
     if isinstance(join_style, str):
```

Comment:
Alternative could be to add a method like 

```
    @classmethod
    def get_value(cls, item):
        try:
            return cls[item].value
        except KeyError:
            raise ValueError("invalid ...")
```

to the Enum definitions (or a base class from which the two enums can inherit), and then here we can just do `cap_style = BufferCapStyles.get_value(cap_style)`. 
Which might give a little bit less code repetition

Reply:
I like the idea of simplifying these via something like a new `ParamEnum` class that provides validation.  One possible disadvantage is that handling this validation in the API function (e.g., `buffer`), we're able to specifically name the parameter in the `ValueError` message, whereas that may not abstract as well to putting validation into the enum.

We're also using the same pattern for STRtree predicates.  However, there, the set of validation options is being used in GeoPandas as per [geopandas/pull/1698](https://github.com/geopandas/geopandas/pull/1698).

Perhaps one path forward would be to pull validation into the enum, and let downstream packages like geopandas use that enum to build their own set of valid options (e.g., via `{e.name for e in BinaryPredicate}` on their end) if they want to handle validation at that level?   (there was a preference for handling this at the downstream level too, for consistency of error messages)

This would have the advantage of exposing only the enums via our public API, but not a corresponding set of values.

/cc @adriangb